settings: make RegisterEnumSetting generic

RaduBerinde opened the pull request:
We change `RegisterEnumSetting` to use any integer type instead of always `int64`. This makes the code safer and cleaner (we no longer need to/from `int64` casts).

Fixes #125486
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125700)
<!-- Reviewable:end -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for cdc. Thanks for the cleanup!

> Reviewed 1 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125700#-:-O-az1UX39pE8NRg9PRr:b-4prwb8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @nkodali, and @rafiss)



<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 15 of 33 files at r1, 1 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125700#-:-O-bAiCgEoqZG5xfpTnH:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @nkodali and @rafiss)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde committed:
settings: make RegisterEnumSetting generic

We change `RegisterEnumSetting` to use any integer type instead of
always `int64`. This makes the code safer and cleaner (we no longer
need to/from `int64` casts).

Epic: none
Release note: None

RaduBerinde commented:
bors r+
